File: test-resource-platform/backend/app/api/v1/resource_inventory.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.auth.authorization import require_roles
from app.auth.dependencies import get_current_user, get_db
from app.identity.models import User, UserRole
from app.leases.service import list_active_machine_users
from app.resources.models import MachineResource, ResourceAdminStatus, ResourcePool
from app.resources.schemas import (
    MachineOccupancyStatus,
    MachineResourceCreateRequest,
    MachineResourcePublic,
    ResourcePoolCreateRequest,
    ResourcePoolPublic,
)
from app.resources.service import (
    create_machine_resource,
    create_resource_pool,
    get_machine_by_resource_code,
    get_resource_pool_by_id,
    get_resource_pool_by_name,
    list_machine_resources,
    list_resource_pools,
    set_machine_admin_status,
    set_resource_pool_active,
)

router = APIRouter()
require_inventory_maintainer = require_roles(UserRole.ADMIN, UserRole.TSE)


@router.get("/resource-pools", response_model=list[ResourcePoolPublic])
def list_pools(
    _: Annotated[User, Depends(get_current_user)],
    session: Annotated[Session, Depends(get_db)],
) -> list[ResourcePool]:
    return list_resource_pools(session)


@router.post(
    "/resource-pools",
    response_model=ResourcePoolPublic,
    status_code=status.HTTP_201_CREATED,
)
def create_pool(
    body: ResourcePoolCreateRequest,
    _: Annotated[User, Depends(require_inventory_maintainer)],
    session: Annotated[Session, Depends(get_db)],
) -> ResourcePool:
    if get_resource_pool_by_name(session, body.name) is not None:
        raise_resource_pool_name_exists()
    try:
        pool = create_resource_pool(session, body)
        session.commit()
    except IntegrityError:
        session.rollback()
        raise_resource_pool_name_exists()
    return pool


@router.get("/machines", response_model=list[MachineResourcePublic])
def list_machines(
    _: Annotated[User, Depends(get_current_user)],
    session: Annotated[Session, Depends(get_db)],
) -> list[MachineResourcePublic]:
    active_machine_users = list_active_machine_users(session)
    machines = []
    for machine in list_machine_resources(session):
        leased_by_username = active_machine_users.get(machine.id)
        machine_public = MachineResourcePublic.model_validate(machine)
        if leased_by_username is not None:
            machine_public = machine_public.model_copy(
                update={
                    "occupancy_status": MachineOccupancyStatus.OCCUPIED,
                    "leased_by_username": leased_by_username,
                }
            )
        machines.append(machine_public)
    session.commit()
    return machines


@router.post("/machines", response_model=MachineResourcePublic, status_code=status.HTTP_201_CREATED)
def create_machine(
    body: MachineResourceCreateRequest,
    _: Annotated[User, Depends(require_inventory_maintainer)],
    session: Annotated[Session, Depends(get_db)],
) -> MachineResource:
    pool = get_resource_pool_by_id(session, body.pool_id)
    if pool is None:
        raise_resource_pool_not_found()
    if not pool.is_active:
        raise_resource_pool_disabled()
    if get_machine_by_resource_code(session, body.resource_code) is not None:
        raise_resource_code_exists()
    try:
        machine = create_machine_resource(session, body)
        session.commit()
    except IntegrityError:
        session.rollback()
        raise_resource_code_exists()
    return machine


@router.post("/resource-pools/{pool_id}/disable", response_model=ResourcePoolPublic)
def disable_pool(
    pool_id: int,
    _: Annotated[User, Depends(require_inventory_maintainer)],
    session: Annotated[Session, Depends(get_db)],
) -> ResourcePool:
    pool = get_resource_pool_by_id(session, pool_id)
    if pool is None:
        raise_resource_pool_not_found()
    set_resource_pool_active(session, pool, False)
    session.commit()
    return pool


@router.post("/resource-pools/{pool_id}/enable", response_model=ResourcePoolPublic)
def enable_pool(
    pool_id: int,
    _: Annotated[User, Depends(require_inventory_maintainer)],
    session: Annotated[Session, Depends(get_db)],
) -> ResourcePool:
    pool = get_resource_pool_by_id(session, pool_id)
    if pool is None:
        raise_resource_pool_not_found()
    set_resource_pool_active(session, pool, True)
    session.commit()
    return pool


@router.post("/machines/{resource_code}/disable", response_model=MachineResourcePublic)
def disable_machine(
    resource_code: str,
    _: Annotated[User, Depends(require_inventory_maintainer)],
    session: Annotated[Session, Depends(get_db)],
) -> MachineResource:
    machine = get_machine_by_resource_code(session, resource_code)
    if machine is None:
        raise_machine_not_found()
    set_machine_admin_status(session, machine, ResourceAdminStatus.DISABLED)
    session.commit()
    return machine


@router.post("/machines/{resource_code}/enable", response_model=MachineResourcePublic)
def enable_machine(
    resource_code: str,
    _: Annotated[User, Depends(require_inventory_maintainer)],
    session: Annotated[Session, Depends(get_db)],
) -> MachineResource:
    machine = get_machine_by_resource_code(session, resource_code)
    if machine is None:
        raise_machine_not_found()
    pool = get_resource_pool_by_id(session, machine.pool_id)
    if pool is None:
        raise_resource_pool_not_found()
    if not pool.is_active:
        raise_resource_pool_disabled()
    set_machine_admin_status(session, machine, ResourceAdminStatus.ACTIVE)
    session.commit()
    return machine


def raise_resource_pool_name_exists() -> None:
    raise HTTPException(
        status_code=status.HTTP_409_CONFLICT,
        detail={
            "error_code": "RESOURCE_POOL_NAME_ALREADY_EXISTS",
            "message": "Resource pool name already exists.",
        },
    )


def raise_resource_pool_not_found() -> None:
    raise HTTPException(
        status_code=status.HTTP_404_NOT_FOUND,
        detail={"error_code": "RESOURCE_POOL_NOT_FOUND", "message": "Resource pool not found."},
    )


def raise_resource_pool_disabled() -> None:
    raise HTTPException(
        status_code=status.HTTP_409_CONFLICT,
        detail={"error_code": "RESOURCE_POOL_DISABLED", "message": "Resource pool is disabled."},
    )


def raise_resource_code_exists() -> None:
    raise HTTPException(
        status_code=status.HTTP_409_CONFLICT,
        detail={
            "error_code": "RESOURCE_CODE_ALREADY_EXISTS",
            "message": "Resource code already exists.",
        },
    )


def raise_machine_not_found() -> None:
    raise HTTPException(
        status_code=status.HTTP_404_NOT_FOUND,
        detail={"error_code": "MACHINE_NOT_FOUND", "message": "Machine not found."},
    )

File: test-resource-platform/frontend/src/api/resourceInventory.ts
import { parseResponse } from '@/api/http';

export type ResourceType = 'PHYSICAL' | 'VIRTUAL';
export type ResourceAdminStatus = 'ACTIVE' | 'MAINTENANCE' | 'DISABLED';
export type ConnectivityStatus = 'UNKNOWN' | 'REACHABLE' | 'UNREACHABLE';
export type MachineOccupancyStatus = 'FREE' | 'OCCUPIED';

export type ResourcePoolPublic = {
  id: number;
  name: string;
  description: string | null;
  location: string | null;
  network_zone: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
};

export type ResourcePoolCreateRequest = {
  name: string;
  description?: string | null;
  location?: string | null;
  network_zone?: string | null;
};

export type MachineResourcePublic = {
  id: number;
  resource_code: string;
  name: string;
  resource_type: ResourceType;
  pool_id: number;
  host_machine_id: number | null;
  admin_status: ResourceAdminStatus;
  connectivity_status: ConnectivityStatus;
  is_critical: boolean;
  owner: string | null;
  architecture: string | null;
  os_name: string | null;
  ip_address: string | null;
  mac_address: string | null;
  bmc_address: string | null;
  tags: string[];
  occupancy_status: MachineOccupancyStatus;
  leased_by_username: string | null;
  active_lease_id: string | null;
  created_at: string;
  updated_at: string;
};

export type MachineResourceCreateRequest = {
  resource_code: string;
  name: string;
  resource_type: ResourceType;
  pool_id: number;
  host_machine_id?: number | null;
  is_critical?: boolean;
  owner?: string | null;
  architecture?: string | null;
  os_name?: string | null;
  ip_address?: string | null;
  mac_address?: string | null;
  bmc_address?: string | null;
  tags?: string[];
};

export async function listResourcePools(): Promise<ResourcePoolPublic[]> {
  const response = await fetch('/api/v1/resource-pools', {
    credentials: 'include'
  });
  return parseResponse<ResourcePoolPublic[]>(response);
}

export async function createResourcePool(
  body: ResourcePoolCreateRequest
): Promise<ResourcePoolPublic> {
  const response = await fetch('/api/v1/resource-pools', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    credentials: 'include',
    body: JSON.stringify(body)
  });
  return parseResponse<ResourcePoolPublic>(response);
}

export async function disableResourcePool(poolId: number): Promise<ResourcePoolPublic> {
  const response = await fetch(`/api/v1/resource-pools/${poolId}/disable`, {
    method: 'POST',
    credentials: 'include'
  });
  return parseResponse<ResourcePoolPublic>(response);
}

export async function enableResourcePool(poolId: number): Promise<ResourcePoolPublic> {
  const response = await fetch(`/api/v1/resource-pools/${poolId}/enable`, {
    method: 'POST',
    credentials: 'include'
  });
  return parseResponse<ResourcePoolPublic>(response);
}

export async function listMachines(): Promise<MachineResourcePublic[]> {
  const response = await fetch('/api/v1/machines', {
    credentials: 'include'
  });
  return parseResponse<MachineResourcePublic[]>(response);
}

export async function createMachine(
  body: MachineResourceCreateRequest
): Promise<MachineResourcePublic> {
  const response = await fetch('/api/v1/machines', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    credentials: 'include',
    body: JSON.stringify(body)
  });
  return parseResponse<MachineResourcePublic>(response);
}

export async function disableMachine(resourceCode: string): Promise<MachineResourcePublic> {
  const response = await fetch(`/api/v1/machines/${resourceCode}/disable`, {
    method: 'POST',
    credentials: 'include'
  });
  return parseResponse<MachineResourcePublic>(response);
}

export async function enableMachine(resourceCode: string): Promise<MachineResourcePublic> {
  const response = await fetch(`/api/v1/machines/${resourceCode}/enable`, {
    method: 'POST',
    credentials: 'include'
  });
  return parseResponse<MachineResourcePublic>(response);
}
